driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+ *
+ */
+@SuppressWarnings("all")
+public final class CodecRegistry {
+
+    private static final Logger logger = LoggerFactory.getLogger(CodecRegistry.class);
+
+    private static final ImmutableSet<TypeCodec<?>> PRIMITIVE_CODECS = ImmutableSet.of(
+        BlobCodec.instance,
+        BooleanCodec.instance,
+        SmallIntCodec.instance,
+        TinyIntCodec.instance,
+        IntCodec.instance,
+        BigintCodec.instance,
+        CounterCodec.instance,
+        DoubleCodec.instance,
+        FloatCodec.instance,
+        VarintCodec.instance,
+        DecimalCodec.instance,
+        VarcharCodec.instance,
+        AsciiCodec.instance,
+        TimestampCodec.instance,
+        DateCodec.instance,
+        TimeCodec.instance,
+        UUIDCodec.instance,
+        TimeUUIDCodec.instance,
+        InetCodec.instance
+    );
+
+    /**
+     * The default set of codecs used by the Java driver.
+     * They contain codecs to handle native types, and collections thereof (lists, sets and maps).
+     *
+     * Note that the default set of codecs has no support for
+     * <a href="https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/marshal/AbstractType.java">Cassandra custom types</a>;
+     * to be able to deserialize values of such types, you need to manually register an appropriate codec.
+     */
+    private static final ImmutableList<TypeCodec<?>> DEFAULT_CODECS;
+
+    static {
+        ImmutableList.Builder<TypeCodec<?>> builder = new ImmutableList.Builder<TypeCodec<?>>();
+        builder.addAll(PRIMITIVE_CODECS);
+        for (TypeCodec<?> primitiveCodec1 : PRIMITIVE_CODECS) {
+            builder.add(new ListCodec(primitiveCodec1));
+            builder.add(new SetCodec(primitiveCodec1));
+            for (TypeCodec<?> primitiveCodec2 : PRIMITIVE_CODECS) {
+                builder.add(new MapCodec(primitiveCodec1, primitiveCodec2));
+            }
+        }
+        DEFAULT_CODECS = builder.build();
+    }
+
+    private static final class CacheKey {
+
+        private final TypeToken<?> javaType;
+
+        private final DataType cqlType;
+
+        public CacheKey(TypeToken<?> javaType, DataType cqlType) {
+            this.javaType = javaType;
+            this.cqlType = cqlType;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+            CacheKey cacheKey = (CacheKey)o;
+            return Objects.equal(javaType, cacheKey.javaType) && Objects.equal(cqlType, cacheKey.cqlType);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(javaType, cqlType);
+        }
+    }
+
+    private final CopyOnWriteArrayList<TypeCodec<?>> codecs;
+
+    private final ConcurrentMap<String, TypeCodec<?>> overrides;
+
+    private final LoadingCache<CacheKey, TypeCodec<?>> cache;
```

Comment:
Overlapping codecs are definitely an advanced feature. The added value is that one can retrieve  a column's value in different ways. e.g. let's suppose a given resultset column contains a JSON string; the following idioms would return the raw string:

```java
row.getString(0); // looks for a codec text <-> String
row.getObject(0, String.class);
```

Wheras the following idioms would return a User instance (assuming the JSON string can be decoded using a registered custom codec):

```java
row.getObject(0); // looks for a codec text <-> ANY - works if the first codec found is the right one - a bit risky
row.getObject(0, User.class); // looks for a codec text <-> User